spring-pulsar-spring-boot-autoconfigure/src/test/java/org/springframework/pulsar/autoconfigure/PulsarReactiveAutoConfigurationTests.java

```diff
@@ -0,0 +1,238 @@
+import org.springframework.boot.test.context.FilteredClassLoader;
+import org.springframework.boot.test.context.assertj.AssertableApplicationContext;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.pulsar.config.PulsarClientFactoryBean;
+import org.springframework.pulsar.core.reactive.DefaultReactivePulsarSenderFactory;
+import org.springframework.pulsar.core.reactive.ReactivePulsarSenderFactory;
+import org.springframework.pulsar.core.reactive.ReactivePulsarSenderTemplate;
+
+/**
+ * Autoconfiguration tests for {@link PulsarReactiveAutoConfiguration}.
+ *
+ * @author Christophe Bornet
+ */
+@SuppressWarnings("unchecked")
+class PulsarReactiveAutoConfigurationTests {
+
+	private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
+			.withConfiguration(AutoConfigurations.of(PulsarAutoConfiguration.class))
+			.withConfiguration(AutoConfigurations.of(PulsarReactiveAutoConfiguration.class));
+
+	@Test
+	void autoConfigurationSkippedWhenReactivePulsarSenderTemplateNotOnClasspath() {
+		this.contextRunner.withClassLoader(new FilteredClassLoader(ReactivePulsarSenderTemplate.class)).run(
+				(context) -> assertThat(context).hasNotFailed().doesNotHaveBean(PulsarReactiveAutoConfiguration.class));
+	}
+
+	@Test
+	void defaultBeansAreAutoConfigured() {
+		this.contextRunner.run((context) -> assertThat(context).hasNotFailed()
+				.hasSingleBean(ReactivePulsarSenderTemplate.class).hasSingleBean(ReactivePulsarClient.class)
+				.hasSingleBean(ProducerCacheProvider.class).hasSingleBean(ReactiveMessageSenderCache.class)
+				.hasSingleBean(ReactivePulsarSenderFactory.class).getBean(ReactivePulsarSenderTemplate.class));
+	}
+
+	@Test
+	void customReactivePulsarSenderTemplateIsRespected() {
+		ReactivePulsarSenderTemplate<String> template = new ReactivePulsarSenderTemplate<>(null);
+		this.contextRunner
+				.withBean("customReactivePulsarSenderTemplate", ReactivePulsarSenderTemplate.class, () -> template)
+				.run((context) -> assertThat(context).hasNotFailed().getBean(ReactivePulsarSenderTemplate.class)
+						.isSameAs(template));
+	}
+
+	@Test
+	void customReactivePulsarClientIsRespected() {
+		ReactivePulsarClient client = AdaptedReactivePulsarClientFactory.create((PulsarClient) null);
+		this.contextRunner.withBean("customReactivePulsarClient", ReactivePulsarClient.class, () -> client).run(
+				(context) -> assertThat(context).hasNotFailed().getBean(ReactivePulsarClient.class).isSameAs(client));
+	}
+
+	@Test
+	void customProducerCacheProviderIsRespected() throws Exception {
+		try (ProducerCacheProvider provider = new ConcurrentHashMapProducerCacheProvider()) {
+			this.contextRunner
+					.withBean("customReactivePulsarSenderTemplate", ProducerCacheProvider.class, () -> provider)
+					.run((context) -> assertThat(context).hasNotFailed().getBean(ProducerCacheProvider.class)
+							.isSameAs(provider));
```

Comment:
It would probably be a good idea (and easy) to add a verification that the PCP provided cache is the one that is ultimately used.

Reply:
We already test below that DefaultReactivePulsarSenderFactory [uses the provided cache](https://github.com/spring-projects-experimental/spring-pulsar/pull/175/files/32cac1e3ee5de637a297dbc494da6a0f12551921#diff-893ec012124f1a540641a8ddbaff726756784d8c2b04d4e2d895e4cf777846b2R150).
Or by ultimately, do you mean it's the one used when calling `ReactivePulsarSenderFactory::createReactiveMessageSender`. That's unit tested in `DefaultReactiveMessageSenderFactoryTests` but not in integration.

Reply:
Right, but we have some [logic in the auto-config](https://github.com/spring-projects-experimental/spring-pulsar/pull/175/files/32cac1e3ee5de637a297dbc494da6a0f12551921#diff-e405d49a17c8d841b416a53a1ffb8848feff7ec3f1feacec13a322811d4cfe55R81) that invokes the provider  that is not tested.

Reply:
We test both cases where the [provider bean is absent](https://github.com/spring-projects-experimental/spring-pulsar/pull/175/files/32cac1e3ee5de637a297dbc494da6a0f12551921#diff-893ec012124f1a540641a8ddbaff726756784d8c2b04d4e2d895e4cf777846b2R214-R215) and where [it is present](https://github.com/spring-projects-experimental/spring-pulsar/pull/175/files/32cac1e3ee5de637a297dbc494da6a0f12551921#diff-893ec012124f1a540641a8ddbaff726756784d8c2b04d4e2d895e4cf777846b2R231-R232).
What is missing ?